File: plugins/plugin-github/rust/src/actions/push_code.rs
#![allow(missing_docs)]

use async_trait::async_trait;
use serde_json::json;

use super::{ActionContext, ActionResult, GitHubAction};
use crate::error::Result;
use crate::types::{CreateCommitParams, FileChange};
use crate::GitHubService;

pub struct PushCodeAction;

#[async_trait]
impl GitHubAction for PushCodeAction {
    fn name(&self) -> &str {
        "PUSH_GITHUB_CODE"
    }

    fn description(&self) -> &str {
        "Creates a commit with file changes and pushes to a GitHub branch."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "COMMIT_CODE",
            "PUSH_CHANGES",
            "COMMIT_FILES",
            "PUSH_FILES",
            "GIT_PUSH",
            "SAVE_CODE",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        Ok(text.contains("push")
            || text.contains("commit")
            || text.contains("save")
            || text.contains("upload"))
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &GitHubService,
    ) -> Result<ActionResult> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let message = context
            .state
            .get("message")
            .and_then(|m| m.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| text.chars().take(100).collect());

        let branch = context
            .state
            .get("branch")
            .and_then(|b| b.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| "main".to_string());

        let files: Vec<FileChange> = context
            .state
            .get("files")
            .and_then(|f| serde_json::from_value(f.clone()).ok())
            .unwrap_or_default();

        if files.is_empty() {
            return Ok(ActionResult::error("No files to commit"));
        }

        let author_name = context
            .state
            .get("authorName")
            .and_then(|a| a.as_str())
            .map(|s| s.to_string());

        let author_email = context
            .state
            .get("authorEmail")
            .and_then(|a| a.as_str())
            .map(|s| s.to_string());

        let params = CreateCommitParams {
            owner: context.owner.clone(),
            repo: context.repo.clone(),
            message,
            files,
            branch,
            parent_sha: None,
            author_name,
            author_email,
        };

        let commit = service.create_commit(params).await?;
        let short_sha = commit.sha.chars().take(7).collect::<String>();

        Ok(ActionResult::success(
            format!("Pushed commit {} to {}", short_sha, &commit.html_url),
            json!({
                "sha": commit.sha,
                "html_url": commit.html_url,
                "message": commit.message,
            }),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_validate_with_push_keyword() {
        let action = PushCodeAction;
        let context = ActionContext {
            message: json!({
                "content": { "text": "Push the changes to the repository" }
            }),
            owner: "test".to_string(),
            repo: "test".to_string(),
            state: json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_with_commit_keyword() {
        let action = PushCodeAction;
        let context = ActionContext {
            message: json!({
                "content": { "text": "Commit these files to the repository" }
            }),
            owner: "test".to_string(),
            repo: "test".to_string(),
            state: json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_without_keywords() {
        let action = PushCodeAction;
        let context = ActionContext {
            message: json!({
                "content": { "text": "Hello world" }
            }),
            owner: "test".to_string(),
            repo: "test".to_string(),
            state: json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[test]
    fn test_action_properties() {
        let action = PushCodeAction;
        assert_eq!(action.name(), "PUSH_GITHUB_CODE");
        assert!(!action.description().is_empty());
        assert!(action.similes().contains(&"COMMIT_CODE"));
        assert!(action.similes().contains(&"PUSH_CHANGES"));
    }
}

File: plugins/plugin-github/typescript/actions/pushCode.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { GITHUB_SERVICE_NAME, type GitHubService } from "../service";
import {
  type CreateCommitParams,
  createCommitSchema,
  type FileChange,
  formatZodErrors,
} from "../types";

const spec = requireActionSpec("PUSH_CODE");

const examples: ActionExample[][] = [
  [
    {
      name: spec.name,
      content: {
        text: "Push the file changes to the feature/dark-mode branch with message 'Add dark mode styles'",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "I'll commit and push those changes to feature/dark-mode.",
        actions: ["PUSH_GITHUB_CODE"],
      },
    },
  ],
  [
    {
      name: "{{user1}}",
      content: {
        text: "Commit these files to main: README.md with content 'Hello World'",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "Committing README.md to main branch.",
        actions: ["PUSH_GITHUB_CODE"],
      },
    },
  ],
];

export const pushCodeAction: Action = {
  name: "PUSH_GITHUB_CODE",
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const service = runtime.getService(GITHUB_SERVICE_NAME);
    if (!service) {
      return false;
    }

    const text = (message.content as Content).text?.toLowerCase() ?? "";
    return (
      text.includes("push") ||
      text.includes("commit") ||
      text.includes("save") ||
      text.includes("upload")
    );
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = runtime.getService<GitHubService>(GITHUB_SERVICE_NAME);

    if (!service) {
      logger.error("GitHub service not available");
      if (callback) {
        await callback({
          text: "GitHub service is not available. Please ensure the plugin is properly configured.",
        });
      }
      return { success: false };
    }

    try {
      const content = message.content as Content;
      const text = content.text ?? "";

      const files = (state?.files as FileChange[]) ?? [];

      const params: CreateCommitParams = {
        owner: (state?.owner as string) ?? service.getConfig().owner ?? "",
        repo: (state?.repo as string) ?? service.getConfig().repo ?? "",
        message: (state?.message as string) ?? text.slice(0, 100),
        files,
        branch: (state?.branch as string) ?? service.getConfig().branch ?? "main",
        authorName: state?.authorName as string | undefined,
        authorEmail: state?.authorEmail as string | undefined,
      };

      const validation = createCommitSchema.safeParse(params);
      if (!validation.success) {
        const errors = formatZodErrors(validation.error);
        logger.error(`Invalid commit parameters: ${errors}`);
        if (callback) {
          await callback({
            text: `I couldn't push the code due to missing information: ${errors}`,
          });
        }
        return { success: false };
      }

      const commit = await service.createCommit(params);

      logger.info(`Created commit ${commit.sha.slice(0, 7)} on ${params.branch}`);

      if (callback) {
        await callback({
          text: `Pushed ${files.length} file(s) to ${params.branch}.\n\nCommit: ${commit.sha.slice(0, 7)}\nMessage: ${commit.message}\n\nView at: ${commit.htmlUrl}`,
        });
      }

      return { success: true };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      logger.error(`Failed to push code: ${errorMessage}`);

      if (callback) {
        await callback({
          text: `Failed to push the code: ${errorMessage}`,
        });
      }

      return { success: false };
    }
  },

  examples,
};

export default pushCodeAction;
